apps/emqx_connector/src/mqtt/emqx_connector_mqtt_worker.erl

```diff
@@ -60,172 +60,252 @@
 
-send_to_remote_async(Pid, Msg, Callback) when is_pid(Pid) ->
-    gen_statem:cast(Pid, {send_to_remote_async, Msg, Callback});
-send_to_remote_async(Name, Msg, Callback) ->
-    gen_statem:cast(name(Name), {send_to_remote_async, Msg, Callback}).
+do_send(Name, {true, Msg}) ->
+    case emqtt:publish(pid(Name), Msg) of
+        ok ->
+            ok;
+        {ok, #{reason_code := RC}} when
+            RC =:= ?RC_SUCCESS;
+            RC =:= ?RC_NO_MATCHING_SUBSCRIBERS
+        ->
+            ok;
+        {ok, #{reason_code := RC, reason_code_name := Reason}} ->
+            ?SLOG(warning, #{
+                msg => "remote_publish_failed",
+                message => Msg,
+                reason_code => RC,
+                reason_code_name => Reason
+            }),
+            {error, Reason};
+        {error, Reason} ->
+            ?SLOG(info, #{
+                msg => "client_failed",
+                reason => Reason
+            }),
+            {error, Reason}
+    end;
+do_send(_Name, false) ->
+    ok.
 
-%% @doc Return all forwards (local subscriptions).
--spec get_forwards(id()) -> [topic()].
-get_forwards(Name) -> gen_statem:call(name(Name), get_forwards, timer:seconds(1000)).
+send_to_remote_async(Name, MsgIn, Callback) ->
+    trycall(
+        fun() -> do_send_async(Name, export_msg(Name, MsgIn), Callback) end,
+        #{badarg => {error, disconnected}}
+    ).
 
-%% @doc Return all subscriptions (subscription over mqtt connection to remote broker).
--spec get_subscriptions(id()) -> [{emqx_types:topic(), qos()}].
-get_subscriptions(Name) -> gen_statem:call(name(Name), get_subscriptions).
+do_send_async(Name, {true, Msg}, Callback) ->
+    emqtt:publish_async(pid(Name), Msg, _Timeout = infinity, Callback);
+do_send_async(_Name, false, _Callback) ->
+    ok.
 
-callback_mode() -> [state_functions].
+ref(Pid) when is_pid(Pid) ->
+    Pid;
+ref(Term) ->
+    ?REF(Term).
 
-%% @doc Config should be a map().
-init(#{name := Name} = ConnectOpts) ->
-    ?SLOG(debug, #{
-        msg => "starting_bridge_worker",
-        name => Name
-    }),
-    erlang:process_flag(trap_exit, true),
-    State = init_state(ConnectOpts),
-    self() ! idle,
-    {ok, idle, State#{
-        connect_opts => pre_process_opts(ConnectOpts)
-    }}.
-
-init_state(Opts) ->
-    ReconnDelayMs = maps:get(reconnect_interval, Opts, ?DEFAULT_RECONNECT_DELAY_MS),
-    StartType = maps:get(start_type, Opts, manual),
-    Mountpoint = maps:get(forward_mountpoint, Opts, undefined),
-    MaxInflightSize = maps:get(max_inflight, Opts, ?DEFAULT_INFLIGHT_SIZE),
-    Name = maps:get(name, Opts, undefined),
-    #{
-        start_type => StartType,
-        reconnect_interval => ReconnDelayMs,
-        mountpoint => format_mountpoint(Mountpoint),
-        max_inflight => MaxInflightSize,
-        connection => undefined,
-        name => Name
-    }.
+trycall(Fun, Else) ->
+    try
+        Fun()
+    catch
+        error:badarg ->
+            maps:get(badarg, Else);
```

Comment:
it seems all `badarg`s are mapped to `disconnected`, maybe simply return `disconnected` when `badarg` happens.